Implement more sophisticated routing

aaronsnig501 opened the pull request:
I'm having way too much fun with this haha

aaronsnig501 committed:
Implement more sophisticated routing

aaronsnig501 committed:
Add Bulma for gh-pages

aaronsnig501 committed:
Changed order of JS file references

aaronsnig501 committed:
Remove implicit call to the home URL

aaronsnig501 committed:
Add BASE_URL to account for redirection

aaronsnig501 committed:
Attempt to parse the window.location in order to facilitate running the project on gh-pages

aaronsnig501 committed:
Fix the BASE_URL issue

aaronsnig501 committed:
Fix the BASE_URL issue

yoniLavi commented:
The routing and the documentation is really nice, thanks Aaron!

But I think we should get rid of Bulma. Other than the annoying attribution requirement, it is licensed as NonCommercial, so we wouldn't be able to use this for Code Institute.

aaronsnig501 commented:
Could still do with adding route params though :P

Agreed on both accounts regarding Bulma haha.

I'll remove it tomorrow evening and replace it with something else.

I'm starting to think that it might be worth introducing templating soon. Controllers don't really have any context without it. What do you think?

yoniLavi commented:
Adding templating and route params definitely sounds good, but I probably won't have time to work on this myself in the next while, so it's up to you :)

aaronsnig501 committed:
Added rudimentary templating engine

aaronsnig501 commented:
I've removed Bulma and added in some very blue CSS (for some reason I couldn't get Bootstrap/Foundation to work for some reason.) I'll look into that soon.

I've got some rudimentary templating in place so using context objects in the controller (something similar to Django.) Each property is then bound to the template. Maybe I'll add something more sophisticated than Regex in the future, but this works for now, and doesn't seem to affect performance too much :)

yoniLavi commented on js/vendor/mini-html5-framework/templater.js:
```diff
@@ -0,0 +1,31 @@
+    });
+
+    function parser(template) {
+        var t, re;
```
Rather than creating a new variable `t`, why not just reuse `template`. It'd be a bit clearer, no?
Or do you have a stylistic aversion to reassigning to parameters?

And concerning `re`, what's the need for it? it shadows the one from outside and isn't actually used here, right?

yoniLavi commented on js/vendor/mini-html5-framework/templater.js:
```diff
@@ -0,0 +1,31 @@
+function render(template, context) {
+
+    var found = {},
+        re = /{{\s*([^}]+)\s*}}/g,
```
Here you accept arbitrary whitespace around the variable, but in `parser` you only handle "{{ " and " }}" . We should be consistent about this.

yoniLavi commented on js/vendor/mini-html5-framework/templater.js:
```diff
@@ -0,0 +1,31 @@
+        }
+    });
+
+    function parser(template) {
```
Just wondering, why did you choose to use a nested function here?
I'm slightly averse to these unless required for specific reasons (e.g. closures or recursion). Is it to make it private?

Generally I'd prefer making it a regular function and passing in `found` and `context` explicitly as parameters, to make the data flow clearer.

yoniLavi commented on js/vendor/mini-html5-framework/templater.js:
```diff
@@ -0,0 +1,31 @@
+        parsedArray = [],
+        curMatch;
+
+    while(curMatch = re.exec(template)) {
```
Ah, funny, I didn't know about this repeated `exec` feature for global regexp's. Perhaps add a comment here to make the use of this feature clear?
